lib/media/media_source_engine.js

```diff
@@ -1359,6 +1359,73 @@ shaka.media.MediaSourceEngine = class {
     return segment;
   }
 
+  /**
+
+   * Prepare the SourceBuffer to parse a potentially new type or codec.
+   *
+   * HACK a suppression due to missing SourceBuffer.changeType definition.
+   * @suppress {missingProperties}
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   * @param {string} newFullMimeType
+   * @private
+   */
+  change_(contentType, newFullMimeType) {
+    shaka.log.debug(`Change Type: ${contentType}`);
+    shaka.log.debug(
+        `${this.sourceBufferTypes_[contentType]} -> ${newFullMimeType}`);
+    if (shaka.media.MediaSourceEngine.isChangeTypeSupported()) {
+      this.sourceBuffers_[contentType].changeType(newFullMimeType);
+    } else {
+      shaka.log.debug('Change Type not supported');
+    }
+    // Fake an 'updateend' event to resolve the operation.
+    this.onUpdateEnd_(contentType);
+  }
+
+  /**
+  * Enqueue an operation to prepare the SourceBuffer to parse a potentially new
+  * type or codec.
+  *
+  * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+  * @param {string} newFullMimeType
+  * @return {!Promise}
+  */
+  changeType(contentType, newFullMimeType) {
+    return this.enqueueOperation_(
+        contentType,
+        // eslint-disable-next-line no-restricted-syntax
+        this.change_.bind(this, contentType, newFullMimeType));
+  }
+
+  /**
+   * Resets the MediaSource
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   * @private
+   */
+  async reset_(contentType, startTime) {
+    this.mediaSourceOpen_ = new shaka.util.PublicPromise();
+    this.mediaSource_ = this.createMediaSource(this.mediaSourceOpen_);
+    await this.mediaSourceOpen_;
+
+    // Fake a seek to catchup the playhead.
+    this.video_.currentTime = startTime;
+
+    // Fake an 'updateend' event to resolve the operation.
+    this.onUpdateEnd_(contentType);
+  }
+
+  /**
+   *
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   */
+  reset(contentType) {
+    return this.enqueueOperation_(
+        contentType,
+        // eslint-disable-next-line no-restricted-syntax
+        this.reset_.bind(this, contentType, this.video_.currentTime),
```

Comment:
Although now that I think about it, the `reset` operation affects both the video and audio, right? Since you're creating a whole new MediaSource. If this is called for the audio, we also want it to prevent new stuff from being appended to the video, that sort of thing.

So... I think this should probably be `enqueueBlockingOperation_` instead of `enqueueOperation_`.

Reply:
Updated to use blocking operation. Refinement included updating the jsdoc for reset and replacing bind with arrow functions in similar fashion as changeType request.